tony-core/src/main/java/com/linkedin/tony/ApplicationMaster.java

```diff
@@ -780,6 +759,147 @@ private ApplicationRpcServer setupRPCService(String hostname) {
     return rpcServer;
   }
 
+  private class TaskScheduler {
+    private Map<TensorFlowContainerRequest, Map<String, Integer>> taskDependencyMap = new HashMap<>();
+    boolean dependencyCheckPassed = true;
+
+    private void scheduleTasks() {
+      session.setResources(yarnConf, hdfsConf, localResources, containerEnv, hdfsClasspath);
+      final List<TensorFlowContainerRequest> requests = session.getContainersRequests();
+
+      if (!isDAG(requests)) {
+        LOG.error("TonY execution graph does not form a DAG, exiting.");
+        session.setFinalStatus(FinalApplicationStatus.FAILED, "TonY task scheduling failed..");
+        dependencyCheckPassed = false;
+        return;
+      }
+
+      // build task dependency graph
+      buildTaskDependency(requests);
+
+      // start/schedule jobs that have no dependency requirements
+      for (TensorFlowContainerRequest request : requests) {
+        if (checkDependencySatisfied(request)) {
+          scheduleJob(request);
+        }
+      }
+    }
+
+    private void buildTaskDependency(List<TensorFlowContainerRequest> requests) {
+      for (TensorFlowContainerRequest request : requests) {
+        for (String dependsOn : request.getDependsOn()) {
+          if (!dependsOn.isEmpty()) {
+            taskDependencyMap.putIfAbsent(request, new HashMap<>());
+            Map<String, Integer> dependenciesForTask = taskDependencyMap.get(request);
+            dependenciesForTask.put(dependsOn, session.getContainerRequestForType(dependsOn).getNumInstances());
+            taskDependencyMap.put(request, dependenciesForTask);
+          }
+        }
+      }
+    }
+
+    private boolean checkDependencySatisfied(TensorFlowContainerRequest request) {
+      return taskDependencyMap.get(request) == null || taskDependencyMap.get(request).isEmpty();
+    }
+
+    private void scheduleJob(TensorFlowContainerRequest request) {
+      AMRMClient.ContainerRequest containerAsk = setupContainerRequestForRM(request);
+      String jobName = request.getJobName();
+      if (!jobTypeToContainerRequestsMap.containsKey(jobName)) {
+        jobTypeToContainerRequestsMap.put(jobName, new ArrayList<>());
+        jobTypeToContainerResources.put(jobName, getContainerResources(jobName));
+      }
+      jobTypeToContainerRequestsMap.get(request.getJobName()).add(containerAsk);
+      for (int i = 0; i < request.getNumInstances(); i++) {
+        amRMClient.addContainerRequest(containerAsk);
+        session.addExpectedTask();
```

Comment:
can we just call the method `setNumExpectedTasks` and do `setNumExpectedTasks(request.getNumInstances())` once outside the forloop?